### PR TITLE
FINERACT-1706: Fix Loan Delinquency Classification after Repayment

### PR DESCRIPTION
**Pull request by josehernandezfintecheandomx**
## Description

The delinquency calculation must to happen as of the latest unpaid chargeback transaction date in order to reflect the right delinquency days and amount

[FINERACT-1706](https://issues.apache.org/jira/browse/FINERACT-1706).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java [hunk 1]
```diff
@@ -109,52 +111,30 @@ public Collection<LoanDelinquencyTagHistoryData> retrieveDelinquencyRangeHistory
...
+            Integer graceDays = 0;
+            if (loan.loanProduct().getLoanProductConfigurableAttributes().getGraceOnArrearsAgingBoolean()) {
+                graceDays = loan.loanProduct().getLoanProductRelatedDetail().getGraceOnArrearsAgeing();
+                if (graceDays == null) {
+                    graceDays = 0;
+                }
+            }
 
-            collectionData.setAvailableDisbursementAmount(loan.getApprovedPrincipal().subtract(loan.getDisbursedAmount()));
+            collectionData.setAvailableDisbursementAmount(loan.getApprovedPrincipal().subtract(loan.getNetDisbursalAmount()));
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java]: i dont think this is correct.

The netDisbursalAmount will be overriden always by the very latest disbursement amount.
**Example:**
- Approved amount is 1000
- I do a disbursement with amount of 400
- The netDisbursalAmount will be 400, so it will set the available disbursement amount is 600
- If I do a 2nd disbursement with amount of 600
- The netDisbursalAmount will be 600, so the available disbursement amount will be set to 400, but that is not correct. We should consider all of the disbursement details to get the correct amount.

@josehernandezfintecheandomx what do you think?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java [hunk 2]
```diff
@@ -109,52 +111,30 @@ public Collection<LoanDelinquencyTagHistoryData> retrieveDelinquencyRangeHistory
...
-                Set<LoanTransactionToRepaymentScheduleMapping> loanTransactionToRepaymentScheduleMappings = loanTransaction
-                        .getLoanTransactionToRepaymentScheduleMappings();
-                LoanTransactionToRepaymentScheduleMapping loanTransactionToRepaymentScheduleMapping = loanTransactionToRepaymentScheduleMappings
-                        .iterator().next();
-                if (!loanTransactionToRepaymentScheduleMapping.getLoanRepaymentScheduleInstallment().isPrincipalCompleted(currency)) {
-                    overdueSinceDate = loanTransaction.getTransactionDate();
-                    break;
-                }
-            }
+            final LocalDate overdueSinceDate = DelinquencyUtils.getOverdueSinceDate(loan, businessDate, graceDays);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java]: You can get the overDueSinceDate and the amount in the same calculation. You dont need to do twice the same:
- DelinquencyUtils.getOverdueSinceDate(loan, businessDate, graceDays) 
- From line 159-177, it is the same.

Please merge them together.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java [hunk 3]
```diff
@@ -174,20 +154,27 @@ public CollectionData calculateLoanCollectionData(final Long loanId) {
             }
 
             // Calculate Delinquency Amount
-            for (LoanRepaymentScheduleInstallment installment : installments) {
+            Map<Integer, BigDecimal> installmentBalances = new HashMap<>();
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java]: This can be removed, the logic in DelinquencyUtils.getOverdueSinceDate can be extended to get the delinquent amount as well.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            }
+        }
+
+        LocalDate overdueSinceDate = null;
+        if (loanRepaymentSchedule != null) {
+            // Default Due date
+            overdueSinceDate = loanRepaymentSchedule.getDueDate();
+        }
+
+        Map<Integer, BigDecimal> installmentBalances = new HashMap<>();
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java]: This is not needed. You already have the first unpaid instalment (see line 42)

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        for (LoanRepaymentScheduleInstallment installment : loan.getRepaymentScheduleInstallments()) {
+            installmentBalances.put(installment.getInstallmentNumber(), installment.getPrincipalCompleted(loanCurrency).getAmount());
+        }
+
+        // If there is some Loan Transaction Chargeback
+        final List<LoanTransaction> loanTransactions = loan.retrieveListOfTransactionsByType(LoanTransactionType.CHARGEBACK);
+        for (LoanTransaction loanTransaction : loanTransactions) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java]: This can be simplified:
- You already knows which is the first unpaid instalment
- Put the total paid principal for that instalment into a temporary variable
- If there are chargeback transactions for that instalment (fetch those loantransactions through the repayment mapping or simply iterate through the loan transactions and check the from- & due date of instalment against the transaction date), as iterating through those transaction, deduct the transaction amount from this temporary variable. And check after the deduction it went to negative. If yes, that transaction is not fully paid yet, so the delinquent amount can be collected from that moment and the overdue since date can be set with the transaction date of that transaction.
BTW question: only chargeback transaction should increase the delinquent amount?

**Review by adamsaghy — CHANGES_REQUESTED** [review]
See my comments!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+                }
+            }
+        }
+
+        if (nextInstallment == null && loan.getRepaymentScheduleInstallments().size() > 0) {
+            nextInstallment = loan.getRepaymentScheduleInstallments().get(loan.getRepaymentScheduleInstallments().size() - 1);
+        }
+
+        // If exists a next no overdue installment
+        if (nextInstallment != null) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java]: I think this logic is not considering properly the chargeback transactions.
**Example:**
- There are 2 installments
- First one is not fully paid and overdue and contains a chargeback transaction
- The second one is not due yet 

The logic here will not set the overdueSinceDate by the chargeback transaction as it was never checked.
The nextInstallment will be set to check the second installment, which is not even due and  might overwrite the overdueSinceDate which is wrong, the first unpaid chargeback transaction is in the first unpaid installment

**Review by adamsaghy — CHANGES_REQUESTED** [review]
See my comments.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+                for (LoanTransactionToRepaymentScheduleMapping mappngInstallment : installment
+                        .getLoanTransactionToRepaymentScheduleMappings()) {
+                    final LoanTransaction loanTransaction = mappngInstallment.getLoanTransaction();
+                    if (loanTransaction.isChargeback() && loanTransaction.getTransactionDate().isBefore(businessDate)) {
+                        amountAvailable = amountAvailable.subtract(loanTransaction.getAmount());
+                        if (amountAvailable.compareTo(BigDecimal.ZERO) < 0 && !overdueSinceDateWasSet) {
+                            overdueSinceDate = loanTransaction.getTransactionDate();
+                            overdueSinceDateWasSet = true;
+                            break;
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java]: We dont want to break the iteration. We need to process all the due chargeback transactions. Please remove the `break`.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java (hunk 1 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java]: @josehernandezfintecheandomx This is still not fixed. Please share your thoughts about the issue I listed above.

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my comments!

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanTransactionChargebackTest.java [hunk 8]
```diff
@@ -160,7 +159,7 @@ public void applyLoanTransactionChargebackInLongTermLoan() {
         }
 
         loanTransactionHelper.evaluateLoanSummaryAdjustments(getLoansLoanIdResponse, Double.valueOf(baseAmount));
-        DelinquencyBucketsHelper.evaluateLoanCollectionData(getLoansLoanIdResponse, 0, Double.valueOf("333.33"));
+        DelinquencyBucketsHelper.evaluateLoanCollectionData(getLoansLoanIdResponse, 0, Double.valueOf("0.00"));
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanTransactionChargebackTest.java:191]: Would you please extend this test case to increase the business date by 1 and run the COB and check the collectiondata will be updated with the proper details?

Comment (josehernandezfintecheandomx) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanTransactionChargebackTest.java:191]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java (hunk 7 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java]: Removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java (hunk 5 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java]: Done

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java (hunk 4 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java]: Removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java (hunk 3 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java]: Removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java (hunk 1 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java]: Updared

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 9]
```diff
@@ -6781,15 +6781,19 @@ public Collection<LoanCharge> getLoanCharges() {
...
-        checkAndFetchLazyCollection(this.loanTransactions);
+        initializeLazyBasicCollections();
         checkAndFetchLazyCollection(this.disbursementDetails);
         checkAndFetchLazyCollection(this.loanTermVariations);
         checkAndFetchLazyCollection(this.collateral);
         checkAndFetchLazyCollection(this.loanOfficerHistory);
         checkAndFetchLazyCollection(this.loanCollateralManagements);
     }
 
+    public void initializeLazyBasicCollections() {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: Please rename this. "Basic collections" is really non-descriptive method name. Suggestion: `fetchRepaymentScheduleAndTransactions()`

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanRepaymentScheduleInstallment.java [hunk 10]
```diff
@@ -140,6 +140,9 @@ public class LoanRepaymentScheduleInstallment extends AbstractAuditableWithUTCDa
     @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY, mappedBy = "installment")
     private Set<LoanInstallmentCharge> installmentCharges = new HashSet<>();
 
+    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.EAGER, mappedBy = "installment")
+    private Set<LoanTransactionToRepaymentScheduleMapping> loanTransactionToRepaymentScheduleMappings = new HashSet<>();
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanRepaymentScheduleInstallment.java:144]: I am a little hesitant whether we need to fetch the mapping eagerly. Can you tell me why did you decide to do so? Eager loading only required if we would read an association when the entity is already detached. I dont think that would happen here (or should)...i guess we need to make sure this entity is used in transactional block

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java [hunk 11]
```diff
@@ -109,89 +103,23 @@ public Collection<LoanDelinquencyTagHistoryData> retrieveDelinquencyRangeHistory
...
     @Override
     public CollectionData calculateLoanCollectionData(final Long loanId) {
-        final LocalDate businessDate = DateUtils.getBusinessLocalDate();
         final Optional<Loan> optLoan = this.loanRepository.findById(loanId);
 
         CollectionData collectionData = CollectionData.template();
         if (optLoan.isPresent()) {
+            final LocalDate businessDate = DateUtils.getBusinessLocalDate();
             final Loan loan = optLoan.get();
+            loan.initializeLazyBasicCollections();
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java]: I would rather put a `@Transactional` annotation on the method then manually try to fetch the lazy associations.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+                                break;
+                            }
+                        }
+                    }
+                }
+            } else if (!firstNotYetDueInstallment) {
+                log.debug("Loan Id: {} with installment {} due date {}", loan.getId(), installment.getInstallmentNumber(),
+                        installment.getDueDate());
+                firstNotYetDueInstallment = true;
+                amountAvailable = installment.getTotalPaid(loanCurrency).getAmount().subtract(installment.getTotalPaidInAdvance());
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java]: Are you sure we need to deduct the total paid in advance?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java (hunk 11 above)
Comment (galovics) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java]: 100% agreed. TBH the initializeLazyXY method calls should be eliminated completely. They are mostly used to overcome the fact that the transaction boundaries are not properly defined.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+import org.apache.fineract.infrastructure.core.service.DateUtils;
+import org.apache.fineract.organisation.monetary.domain.MonetaryCurrency;
+import org.apache.fineract.portfolio.loanaccount.data.CollectionData;
+import org.apache.fineract.portfolio.loanaccount.domain.Loan;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanRepaymentScheduleInstallment;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanTransaction;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanTransactionToRepaymentScheduleMapping;
+
+@Slf4j
+public final class DelinquencyUtils {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java]: I don't like this. When I see the following names:
- Manager
- Utils
- Util
- etc

I'm always careful because usually it's something fishy. Not always but most of the times. I think this is rather a poor implementation for separation of concerns. On one hand, let's forbid ourselves from using static methods unless there's a good reason for it. 

In this case this class is handling business logic. That's not a Util and that shouldn't be handled in static methods.

Try to generalize the class/method around single responsibilities and try to split it that way into components.

Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java]: Bonus, why don't I see unit tests for this class?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 9 above)
Comment (galovics) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: Let's get rid of this method rather. It's spreading a bad pattern across the system. Let's try to force ourselves to properly define our transaction boundaries and the methods like this will never be used.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanRepaymentScheduleInstallment.java (hunk 10 above)
Comment (galovics) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanRepaymentScheduleInstallment.java:144]: > we would read an association when the entity is already detached.

Not necessarily. Eager fetching comes handy when the parent entity - from a business point of view - can't live without the child entity. 
I'm not sure if this is the case here though so I'm also waiting for some explanation here.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanRepaymentScheduleInstallment.java [hunk 14]
```diff
@@ -718,6 +721,9 @@ public void addToCredits(final BigDecimal amount) {
     }
 
     public BigDecimal getTotalPaidInAdvance() {
+        if (this.totalPaidInAdvance == null) {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanRepaymentScheduleInstallment.java]: This is kinda fishy to me again. 
If the column can be null, that means there was no payment in advance, right? We introduce that logic and the users of this class will never be able to make that decision, right? Since we hide it here with this condition.

Also, if this logic has been introduced here, did you check the whole codebase for native SQLs/JPQLs/etc where the same DB table has been used and the same column has been retrieved? We'll need this same behavior there, right?

**Review by galovics — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 9 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: Done, removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanRepaymentScheduleInstallment.java (hunk 10 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanRepaymentScheduleInstallment.java:144]: Done, moved to LAZY

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java (hunk 11 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java]: Done, removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java (hunk 12 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java]: Removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanRepaymentScheduleInstallment.java (hunk 14 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanRepaymentScheduleInstallment.java]: Removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java (hunk 13 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java]: It is now in a service, I can't put in the current Read service because It generates circular dependency

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java [hunk 15]
```diff
@@ -109,89 +104,22 @@ public Collection<LoanDelinquencyTagHistoryData> retrieveDelinquencyRangeHistory
...
     public CollectionData calculateLoanCollectionData(final Long loanId) {
-        final LocalDate businessDate = DateUtils.getBusinessLocalDate();
         final Optional<Loan> optLoan = this.loanRepository.findById(loanId);
 
         CollectionData collectionData = CollectionData.template();
         if (optLoan.isPresent()) {
+            final LocalDate businessDate = DateUtils.getBusinessLocalDate();
             final Loan loan = optLoan.get();
 
+            collectionData = this.loanCollectionDomainService.getOverdueCollectionData(loan, businessDate);
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java]: `this.` not needed.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java [hunk 16]
```diff
@@ -152,61 +149,9 @@ public LoanScheduleDelinquencyData calculateDelinquencyData(LoanScheduleDelinque
...
-    private Long calculateOverdueDays(final LocalDate businessDate, LocalDate overdueSinceDate) {
-        Long overdueDays = 0L;
-        if (overdueSinceDate != null) {
-            overdueDays = DateUtils.getDifferenceInDays(overdueSinceDate, businessDate);
-            if (overdueDays < 0) {
-                overdueDays = 0L;
-            }
-        }
-        return overdueDays;
+        final CollectionData collectionData = this.loanCollectionDomainService.getOverdueCollectionData(loan, businessDate);
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: `this.` not needed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java [hunk 17]
```diff
@@ -152,61 +149,9 @@ public LoanScheduleDelinquencyData calculateDelinquencyData(LoanScheduleDelinque
...
+        log.debug("Delinquency {}", collectionData.toString());
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: No need for the explicit toString call.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java [hunk 18]
```diff
@@ -217,12 +162,8 @@ public CommandProcessingResult applyDelinquencyTagToLoan(Long loanId, JsonComman
         final DelinquencyBucket delinquencyBucket = loan.getLoanProduct().getDelinquencyBucket();
         if (delinquencyBucket != null) {
             final LocalDate businessDate = DateUtils.getBusinessLocalDate();
-
-            final Integer graceOnArrearAgeing = loan.getLoanProduct().getLoanProductRelatedDetail().getGraceOnArrearsAgeing();
-            final LocalDate overdueSinceDate = getOverdueSinceDate(loan, businessDate, graceOnArrearAgeing);
-
-            final Long overdueDays = calculateOverdueDays(businessDate, overdueSinceDate);
-            changes = lookUpDelinquencyRange(loan, delinquencyBucket, overdueDays);
+            final CollectionData collectionData = this.loanCollectionDomainService.getOverdueCollectionData(loan, businessDate);
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: No need for `this.`

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java [hunk 19]
```diff
@@ -233,15 +174,9 @@ public void applyDelinquencyTagToLoan(LoanScheduleDelinquencyData loanDelinquenc
...
-            final Integer graceOnArrearAgeing = loan.getLoanProduct().getLoanProductRelatedDetail().getGraceOnArrearsAgeing();
-
-            LocalDate overdueSinceDate = loanDelinquencyData.getOverdueSinceDate();
-            if (overdueSinceDate == null) {
-                getOverdueSinceDate(loan, businessDate, graceOnArrearAgeing);
-            }
-
-            final Long overdueDays = calculateOverdueDays(businessDate, overdueSinceDate);
-            lookUpDelinquencyRange(loan, delinquencyBucket, overdueDays);
+            final CollectionData collectionData = this.loanCollectionDomainService.getOverdueCollectionData(loan, businessDate);
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: As above.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java [hunk 20]
```diff
@@ -233,15 +174,9 @@ public void applyDelinquencyTagToLoan(LoanScheduleDelinquencyData loanDelinquenc
...
+            log.debug("Delinquency {}", collectionData.toString());
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: No need for toString

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/LoanCollectionDomainService.java [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+public interface LoanCollectionDomainService {
+
+    /**
+     * This method is to calculate the Overdue date and other properties, If the loan is overdue or If there is some
+     * Charge back transaction
+     *
+     * @param loan
+     * @param businessDate
+     */
+    CollectionData getOverdueCollectionData(Loan loan, LocalDate businessDate);
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/LoanCollectionDomainService.java]: If we're always passing the current business date, why is it a parameter? Can't we just resolve the current business date from within the implementation?

**Review by adamsaghy — DISMISSED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanEventApiJsonValidator.java [hunk 22]
```diff
@@ -237,7 +237,7 @@ public void validateChargebackTransaction(final String json) {
         final BigDecimal transactionAmount = this.fromApiJsonHelper
                 .extractBigDecimalWithLocaleNamed(LoanApiConstants.TRANSACTION_AMOUNT_PARAMNAME, element);
         baseDataValidator.reset().parameter(LoanApiConstants.TRANSACTION_AMOUNT_PARAMNAME).value(transactionAmount).notNull()
-                .zeroOrPositiveAmount();
+                .positiveAmount();
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanEventApiJsonValidator.java]: This is a separate bugfix. Please extract it to a different PR so it can be reviewed separately along with its tests.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/LoanCollectionDomainServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,135 @@
...
+import org.apache.fineract.portfolio.loanaccount.data.CollectionData;
+import org.apache.fineract.portfolio.loanaccount.domain.Loan;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanRepaymentScheduleInstallment;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanTransaction;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanTransactionToRepaymentScheduleMapping;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+public class LoanCollectionDomainServiceImpl implements LoanCollectionDomainService {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/LoanCollectionDomainServiceImpl.java]: Don't like the naming still, especially the terminology CollectionData because it doesn't reflect anything about the data it represents.
But I can live with it for now.

The other thing though which I can't live without is proper unit test coverage for this class.

**Review by galovics — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java (hunk 15 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyReadPlatformServiceImpl.java]: Removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 16 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: Removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 17 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: Removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 18 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: Removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 19 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: Removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 20 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: Removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanEventApiJsonValidator.java (hunk 22 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanEventApiJsonValidator.java]: Removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/LoanCollectionDomainService.java (hunk 21 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/LoanCollectionDomainService.java]: Updated

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java (hunk 13 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java]: Done in class `LoanDelinquencyDomainServiceTest`

**Review by galovics — APPROVED** [review]

**Review by adamsaghy — APPROVED** [review]
LGTM
